added scripts for visualizing and calculating metrics for c1t develop…

john-chrosniak opened the pull request:
# PR Details
## Description

This PR includes scripts for visualizing and computing metrics for C1T

## Related Jira Key

[CF-846](https://usdot-carma.atlassian.net/browse/CF-846)

## Motivation and Context

These visualizations will help with debugging and showing progress made throughout C1T development. The metrics will also help quantify the impact of improvements.

## How Has This Been Tested?

The scripts have been tested on bags recorded from the C1T trucks.

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:


- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.


[CF-846]: https://usdot-carma.atlassian.net/browse/CF-846?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

john-chrosniak commented:
![calc_route_deviation](https://github.com/usdot-fhwa-stol/carma-analytics-fotda/assets/43456793/83016c1d-575a-4413-b0bd-d1a7f546946f)
![plot_route_driven](https://github.com/usdot-fhwa-stol/carma-analytics-fotda/assets/43456793/0b9ed9d9-b2c7-4301-93fb-419c287b9cc8)
![plot_steering_rate](https://github.com/usdot-fhwa-stol/carma-analytics-fotda/assets/43456793/e6e3d57f-b1f9-4c67-8126-ea0c08764f2a)

willjohnsonk commented on src/carma-1-tenth/calc_route_deviation.py:
```diff
@@ -0,0 +1,123 @@
+from scipy.interpolate import make_interp_spline
+import os
+
+STEERING_TO_SERVO_OFFSET = 0.425
```
Consider adding a comment to describe where these values come from (params file) and to indicate that they should be updated to fit the parameter values of ackermann_to_vesc

john-chrosniak replied:
comment added in plot_steering_rate.py, this was removed here because it was unused

willjohnsonk commented on src/carma-1-tenth/calc_route_deviation.py:
```diff
@@ -0,0 +1,123 @@
+    reader.set_filter(filt)
+    route_graph = None
+    odom_count = 0
+    odometry = np.zeros((topic_count_dict[odom_topic], 2))
```
does this still register a 0,0 point for the route marker points?

willjohnsonk commented on src/carma-1-tenth/calc_route_deviation.py:
```diff
@@ -0,0 +1,123 @@
+    route_y_points = y_spline(samples)
+    route_deviations = []
+    odom_movements = []
+    for i in range(1, len(odometry)):
```
Seeing as there's a lot of sequential for loops here I'd split them up with comments that briefly say what each is doing

john-chrosniak replied:
Added some comments for each code block

willjohnsonk commented on src/carma-1-tenth/calc_route_deviation.py:
```diff
@@ -0,0 +1,123 @@
+    argdict : dict = vars(args)
+    calc_route_deviation(os.path.normpath(os.path.abspath(argdict["bag_in"])))
+    plt.xlabel("Index")
+    plt.ylabel("Deviation from Route (m)")
```
Consider "Absolute Deviation from Route" for clarity since we aren't reflecting left-right deviations with signed values

john-chrosniak replied:
Changed to absolution route deviation

willjohnsonk commented on src/carma-1-tenth/plot_route_driven.py:
```diff
@@ -0,0 +1,96 @@
+import os
+
+STEERING_TO_SERVO_OFFSET = 0.425
+STEERING_TO_SERVO_GAIN = -0.55
```
See above

willjohnsonk replied:
Also these aren't used in the two previous files, only in the last script

willjohnsonk commented on src/carma-1-tenth/plot_route_driven.py:
Do these all need to be separate files? I'm seeing a lot of function overlap for handling rosbags.

john-chrosniak replied:
Good point, I migrated the rosbag parsing functions to rosbag_utils.py

willjohnsonk commented on src/carma-1-tenth/plot_steering_rate.py:
```diff
@@ -0,0 +1,97 @@
+    argdict : dict = vars(args)
+    plot_steering_rate(os.path.normpath(os.path.abspath(argdict["bag_in"])), "Unfiltered Steering")
+    if argdict["bag_in_2"]:
+        plot_steering_rate(os.path.normpath(os.path.abspath(argdict["bag_in_2"])), "Filtered Steering", 6.7)
```
hardcoded offset should be a variable

john-chrosniak replied:
True, I made that an arg that can be passed in

willjohnsonk commented on src/carma-1-tenth/plot_route_driven.py:
```diff
@@ -0,0 +1,96 @@
+    args = parser.parse_args()
+    argdict : dict = vars(args)
+    plot_route_driven(os.path.normpath(os.path.abspath(argdict["bag_in"])))
+    plt.xlim([-3, 5])
```
consider adding variables instead of hardcoded plot limits

john-chrosniak replied:
I adjusted the plot to autoscale based on the routes coordinates

willjohnsonk commented on src/carma-1-tenth/calc_route_deviation.py:
```diff
@@ -0,0 +1,123 @@
+STEERING_TO_SERVO_OFFSET = 0.425
+STEERING_TO_SERVO_GAIN = -0.55
+
+def get_rosbag_options(path, serialization_format="cdr", storage_id="sqlite3"):
```
using db3 rosbag, will likely want to transition to mcap but likely out of scope

john-chrosniak replied:
the script grabs the storage_id from the rosbags metadata.yaml, so this should work for mcaps as well

adev4a commented on src/carma-1-tenth/plot_route_driven.py:
```diff
@@ -0,0 +1,62 @@
+            msg_type = type_map[topic]
+            msg_type_full = get_message(msg_type)
+            msg = deserialize_message(data, msg_type_full)
+            if msg_type == "visualization_msgs/msg/MarkerArray":
```
Is there a reason we're checking for message type and not topic?

john-chrosniak replied:
Good point, just changed this to check by topic instead

adev4a commented on src/carma-1-tenth/plot_route_driven.py:
```diff
@@ -0,0 +1,62 @@
+if __name__=="__main__":
+    import argparse, argcomplete
+    parser = argparse.ArgumentParser(description="Plot steering rate of C1T trucks")
+    parser.add_argument("bag_in", type=str, help="Bag to load")
```
This argument looks like it is the bag directory instead of the bag file itself

john-chrosniak replied:
Correct this should be the bag directory

adev4a replied:
Can you update the description in the help text to indicate this?

adev4a replied:
Can you update the text to indicate this?

john-chrosniak replied:
done!

adev4a commented:
> Here are the plots generated with each script, feel free to comment on possible improvements!
> 

For plot 1, can we update the x label? Its not very clear what index is representing.
For plot 2, what are the red dots? Can we add a legend to this plot? 
Can we also add titles to all of these plots?

john-chrosniak commented on src/carma-1-tenth/calc_absolute_route_deviation.py:
```diff
@@ -0,0 +1,120 @@
+    topic_metadata_map = {topic_types[i].name: topic_types[i] for i in range(len(topic_types))}
+    return topic_types, type_map, topic_metadata_map, reader
+
+def open_bagfile_writer(filepath: str, serialization_format="cdr", storage_id="sqlite3"):
```
Yeah we aren't writing bag files so I removed this

john-chrosniak commented:
> > Here are the plots generated with each script, feel free to comment on possible improvements!
> 
> For plot 1, can we update the x label? Its not very clear what index is representing.

I changed this to time in seconds to make this more clear.

>  For plot 2, what are the red dots? Can we add a legend to this plot? Can we also add titles to all of these plots?

The red dots are the route we intend to follow and the blue line is the actual path driven. I added a legend to indicate this and added titles as well.

john-chrosniak commented:
![plot_steering_rate](https://github.com/usdot-fhwa-stol/carma-analytics-fotda/assets/43456793/176b92ef-5453-49dc-9da7-0978ff137431)
![plot_route_driven](https://github.com/usdot-fhwa-stol/carma-analytics-fotda/assets/43456793/8d62af6a-f144-444d-a987-088e8d96aa69)
![plot_absolute_deviation_from_route](https://github.com/usdot-fhwa-stol/carma-analytics-fotda/assets/43456793/08e8301d-a7b5-4db0-8855-7ff5b05a3b81)

john-chrosniak commented:
![updated_route_deviation](https://github.com/usdot-fhwa-stol/carma-analytics-fotda/assets/43456793/b9ba3ad5-db4f-4177-9790-6dd4b48dfc78)

Changed x-axis of route deviation plot to downtrack

adev4a commented on src/carma-1-tenth/plot_steering_rate.py:
```diff
@@ -0,0 +1,67 @@
+
+
+if __name__=="__main__":
+    import argparse, argcomplete
```
Can we add all imports to the top?

john-chrosniak replied:
done!

adev4a commented on src/carma-1-tenth/plot_steering_rate.py:
```diff
@@ -0,0 +1,67 @@
+    parser.add_argument("--filtered_bag_offset", type=float, default=0.0, help="Time offset for start of bag with filtered steering")
+    parser.add_argument("--unfiltered_bag", type=str, help="Bag with unfiltered steering")
+    parser.add_argument("--unfiltered_bag_offset", type=float, default=0.0, help="Time offset for start of bag with unfiltered steering")
+    parser.add_argument("--png_out", type=str, help="Output file")
```
Is this like a boolean flag? Can you update the comment in the help text to something like "will output a png file if flag enabled"

john-chrosniak replied:
This command is a filepath to save the plot as a png, I updated the comment to indicate this

adev4a commented on src/carma-1-tenth/plot_absolute_route_deviation.py:
```diff
@@ -0,0 +1,103 @@
+            msg_type = type_map[topic]
+            msg_type_full = get_message(msg_type)
+            msg = deserialize_message(data, msg_type_full)
+            if msg_type == "visualization_msgs/msg/MarkerArray":
```
Can we replace this with the topic name as well?

john-chrosniak replied:
done!

adev4a commented:
Few small comments but this looks great overall. Thanks for updating the plots as well. One additional note - can you add a comment in each of the scripts about what they are doing? Just so that anyone looking through the script can figure out what they are meant for

john-chrosniak commented:
![crosstrack_vs_downtrack](https://github.com/usdot-fhwa-stol/carma-analytics-fotda/assets/43456793/306788f6-a3ca-46e0-9a9b-37d8b53097ab)
 
Updated plot
